### PR TITLE
Make sure we don't clean up an active handle during a next download.

### PR DESCRIPTION
**Pull request by maleadt**
Fixes https://github.com/JuliaLang/Downloads.jl/issues/99

### CODE REVIEW

**Comment by giordano** [comment]
Is any of the tests in #99 reproducible in CI?

**Comment by maleadt** [comment]
Yes, I've added the MWE as a test.

**Comment by fredrikekre** [comment]
Does it also fix #95 or are they completely separate issues? Have you tried enabling HTTP/2 wich this patch?

**Comment by maleadt** [comment]
> Does it also fix #95 or are they completely separate issues? Have you tried enabling HTTP/2 wich this patch?

That seems very possible, but I couldn't reproduce with HTTP 2.0 re-enabled (and this patch either disabled or enabled). Could you try on your system?

File: src/Curl/Multi.jl [hunk 1]
```diff
@@ -43,7 +43,13 @@ end
 function add_handle(multi::Multi, easy::Easy)
     lock(multi.lock) do
         isempty(multi.easies) && preserve_handle(multi)
-        multi.handle == C_NULL && init!(multi)
+        if multi.handle == C_NULL
+            init!(multi)
+        else
+            # if reusing a multi, it's possible we had queued up a cleanup of the handle.
+            # stop that timer so that we don't accidentaly clean up during another download.
+            uv_timer_stop(multi.timer)
```
Comment (StefanKarpinski) [inline src/Curl/Multi.jl:51]: Seems better to just hoist the `uv_timer_stop(multi.timer)` call out of `init!`.

Comment (StefanKarpinski) [reply src/Curl/Multi.jl:51]: I.e. something like this:
```diff
diff --git a/src/Curl/Multi.jl b/src/Curl/Multi.jl
index f16e365..2a5f538 100644
--- a/src/Curl/Multi.jl
+++ b/src/Curl/Multi.jl
@@ -18,7 +18,6 @@ mutable struct Multi
 end

 function init!(multi::Multi)
-    uv_timer_stop(multi.timer)
     multi.handle = curl_multi_init()
     add_callbacks(multi)
     set_defaults(multi)
@@ -43,7 +42,12 @@ end
 function add_handle(multi::Multi, easy::Easy)
     lock(multi.lock) do
         isempty(multi.easies) && preserve_handle(multi)
-        multi.handle == C_NULL && init!(multi)
+        if multi.handle == C_NULL
+            init!(multi)
+        else
+            # stop grace timer
+            uv_timer_stop(multi.timer)
+        end
         push!(multi.easies, easy)
         @check curl_multi_add_handle(multi.handle, easy.handle)
     end
```
The remaining question is whether the call to `uv_timer_stop` should happen on every added handle or only the first one (when `multi.handle == C_NULL`). I suspect the latter, but I'm not 100% sure. If so, then this code was calling `uv_timer_stop` exactly when it shouldn't and not calling it when it should 😬

**Pull request closed without merging**